### PR TITLE
Journal: Use sole_replica()

### PR DESCRIPTION
**Pull request by sentientwaffle**
## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

### CODE REVIEW

File: src/vsr/journal.zig [hunk 1]
```diff
@@ -1643,7 +1643,7 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             assert(journal.has_dirty(message.header));
 
             const write = journal.writes.acquire() orelse {
```
Comment (matklad) [inline src/vsr/journal.zig:1645]: Just for my understanding:

- it indeed can happen that a node runs out of IOPs, so prepare won't get written
- however, the prepare would still be in memory (pipeline or cache), so we'll retry writing it

?

**Review by matklad — APPROVED** [review]
LGTM!

File: src/vsr/journal.zig (hunk 1 above)
Comment (sentientwaffle) [reply src/vsr/journal.zig:1645]: > it indeed can happen that a node runs out of IOPs, so prepare won't get written

Yes

> however, the prepare would still be in memory (pipeline or cache), so we'll retry writing it

Depends if we're a primary or backup:

- If we're a primary and this is not a repair, then yes, it is in the pipeline, and we will try again.
- If we're a primary and this is a repair, then the retry will be driven by us requesting the prepare all over again.
- If we're a backup, whether it is a repair or not, retries are driven by requesting prepares. We do check our cache for prepares before requesting them.

File: src/vsr/journal.zig [hunk 2]
```diff
@@ -1643,7 +1643,7 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
...
-                assert(replica.replica_count > 1);
+                assert(!replica.sole_replica());
```
Comment (jorangreef) [inline src/vsr/journal.zig:1646]: Perhaps `replica.solo()` would read better?

Comment (matklad) [reply src/vsr/journal.zig:1646]: 🤔 The part about *replica* is important semantically: sole_replica might still be a member of cluster with several standbys.

otoh, `solo` is quite catchy, and names the whole concept! I think on balance it’s better, will try rename

**Pull request merged**